### PR TITLE
WIP Profile Sidebar

### LINKED ISSUE #208: Profile Sidebar: Public vs. Private
This issue is about adding the "public" and "private" sections here.

image

Active/inactive style:

image

For a first pass, simplest solution probably is to leave the profile backend and query as-is, and just filter on the client side what is shown.

The state of what section is open should be a part of the URL, possibly a new parameter like `?section=public`

### PR DESCRIPTION
**Pull request by curran**
Noticed this branch from @milafrerichs 

Closes https://github.com/datavis-tech/vizhub-issue-tracker/issues/208

### CODE REVIEW

File: packages/neoFrontend/package.json [hunk 1]
```diff
@@ -46,6 +46,7 @@
     "not op_mini all"
   ],
   "devDependencies": {
+    "@testing-library/react": "^10.3.0",
```
Comment (curran) [inline packages/neoFrontend/package.json:54]: I'm curious, why not just use built-in Jest setup? https://create-react-app.dev/docs/running-tests/

File: packages/neoFrontend/src/pages/ProfilePage/Body/index.js [hunk 2]
```diff
@@ -2,18 +2,23 @@ import React, { useContext } from 'react';
...
 import { VizPreviews, VizPreview } from '../../../VizPreview';
-import { Wrapper, Content } from '../../styles';
+import { Wrapper, Content, Sidebar } from '../../styles';
 
 export const Body = () => {
   const profilePageData = useContext(ProfilePageDataContext);
   const { user, visualizationInfos } = profilePageData;
 
+  const privateViz = visualizationInfos.filter((d) => { return d.privacy === "private"});
+  const publicViz = visualizationInfos.filter((d) => { return d.privacy === "public"});
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/Body/index.js]: So here, I recommend to use an accessor, since in reality a value of `undefined` for `privacy` translates to `"public"`. See `packages/presenters/src/accessors.js`, `getVizPrivacy`.

File: packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    <Themed>
+      <AuthContext.Provider value={me}>
+        <ProfilePageDataContext.Provider value={profilePageData}>
+          <ProfilePage />
+        </ProfilePageDataContext.Provider>
+      </AuthContext.Provider>
+    </Themed>
+  )
+  const { getByText } = render(tree, { wrapper: MemoryRouter })
+  expect(getByText('My little private visualization')).toBeInTheDocument()
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js]: Nice!

**Review by curran — COMMENTED** [review]
@milafrerichs I like the direction here. Please keep going!

**Comment by curran** [comment]
@milafrerichs Looking forward to see this profile page sidebar feature land! Let me know if you are blocked by anything.

**Comment by curran** [comment]
@milafrerichs Note the merge conflicts. A lot has change since this branch was created. May be easier to branch from current master and move the work into a new PR :)

File: packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js [hunk 4]
```diff
@@ -1,17 +1,19 @@
 import '@testing-library/jest-dom/extend-expect';
+import { screen } from '@testing-library/dom'
 import React from 'react';
-import { render } from '@testing-library/react'
+import { render, fireEvent } from '@testing-library/react'
 import { MemoryRouter } from 'react-router-dom'
 import { AuthContext } from '../../authentication';
 
-import { ProfilePage } from './index.js';
+import { Body as ProfilePage } from './Body';
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js:9]: My personal preference is not to rename on import, but instead to rename at the source, e.g. `ProfilePageBody`, so that the same thing is not referred to using two names (which I find confusing). Thanks!

File: packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js [hunk 5]
```diff
@@ -23,7 +25,28 @@ test('NameConsumer shows value from provider', () => {
       </AuthContext.Provider>
     </Themed>
   )
-  const { getByText } = render(tree, { wrapper: MemoryRouter })
-  expect(getByText('My little private visualization')).toBeInTheDocument()
-  expect(getByText('My big public visualization')).toBeInTheDocument()
+  render(tree, { wrapper: MemoryRouter })
+  expect(screen.getByText('My little private visualization')).toBeInTheDocument()
+  expect(screen.getByText('My big public visualization')).toBeInTheDocument()
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js:34]: This is a really cool style of test. This sets a great precedent for future tests of this nature.

**Comment by milafrerichs** [comment]
> @milafrerichs Note the merge conflicts. A lot has change since this branch was created. May be easier to branch from current master and move the work into a new PR :)

@curran
the merge conflicts are not that crucial. 
it's more the design. 
any news on that front? since this was the profile page but the designs do not include this specific thing (it was/is in the "dashboard").

File: packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js (hunk 4 above)
Comment (milafrerichs) [reply packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js:9]: ok, sure, will do that. it was only for the tests that's why I thought this might not be an issue, I could use `Body` but that was too ambiguous :)

File: packages/neoFrontend/src/pages/ProfilePage/Body/index.js [hunk 6]
```diff
@@ -1,22 +1,45 @@
...
+
 
 export const Body = () => {
   const profilePageData = useContext(ProfilePageDataContext);
   const { user, visualizationInfos } = profilePageData;
 
+  const [visualizations, setVisualizations] = useState(visualizationInfos);
+
+  function showPublic(e) {
+    setVisualizations(visualizationInfos.filter((d) => { return getVizPrivacy(d) !== "private"}));
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/Body/index.js]: The accessor `getVizPrivacy` expects a `Visualization`, but you're passing a `VizualizationInfo`, so it breaks. Suggestion: create a `getVizInfoPrivacy` accessor that expects this sort of object, and refactor `getVizPrivacy` to use it.

File: packages/neoFrontend/src/pages/ProfilePage/Body/index.js [hunk 7]
```diff
@@ -1,22 +1,45 @@
...
+  }
+  function showPrivate(e) {
+    setVisualizations(visualizationInfos.filter((d) => { return getVizPrivacy(d) === "private"}));
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/Body/index.js]: Also a minor thing, the syntax could be a bit cleaner.

```suggestion
    setVisualizations(visualizationInfos.filter((d) => getVizInfoPrivacy(d) === "private"));
```

File: packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+          <ProfilePage />
+        </ProfilePageDataContext.Provider>
+      </AuthContext.Provider>
+    </Themed>
+  )
+  render(tree, { wrapper: MemoryRouter })
+  expect(screen.getByText('My little private visualization')).toBeInTheDocument()
+  expect(screen.getByText('My big public visualization')).toBeInTheDocument()
+})
+test('clicking private shows private viz', () => {
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js:36]: Missing a test for clicking public viz.

Also, I'm surprised that this test passes, because when I run it locally and click on private, it crashes due to the mismatched accessor.

File: packages/neoFrontend/src/pages/VizPage/LinkWithIcon/index.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import { Wrapper, Link, LinkText, Blank } from './styles';
+
+export const LinkWithIcon = ({ onClick, icon, title, children }) => {
+  const [isHovering, setIsHovering] = useState(false);
+  const onMouseEnter = useCallback(() => setIsHovering(true), []);
+  const onMouseLeave = useCallback(() => setIsHovering(false), []);
+  const Icon = svgs[icon] || Blank;
+  return (
+    <Wrapper>
+      <Link
```
Comment (curran) [inline packages/neoFrontend/src/pages/VizPage/LinkWithIcon/index.js:9]: The currently active link should be styled with an active style, as specified in the [original issue](https://github.com/datavis-tech/vizhub-issue-tracker/issues/208).

image

Apologies if the intention of this was not clear. The idea is that if you're currently viewing "Public", then that link should have this active style with the shaded background. If you're currently viewing "Private", then that link should have this active style. The style is _not_ intended to represent the hovered link, but rather the active link.

File: packages/neoFrontend/src/pages/ProfilePage/styles.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+export const Main = styled.div`
+  display: flex;
+  width: 85%;
+`;
+
+export const Sidebar = styled.div`
+  display: flex;
+  flex-direction: column;
+  width: 15%;
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/styles.js]: Please make the sidebar a fixed number of pixels, and allow the `Main` part to expand with `flex:1`. This current setup does not work well for small width.

image

Also, can you please add `6px` padding, so that it aligns with the VizHub logo when the browser is resized to be small. Thanks!

File: packages/neoFrontend/src/pages/VizPage/LinkWithIcon/index.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import React, { useState, useCallback } from 'react';
+import * as svgs from '../../../svg';
+import { Wrapper, Link, LinkText, Blank } from './styles';
+
+export const LinkWithIcon = ({ onClick, icon, title, children }) => {
+  const [isHovering, setIsHovering] = useState(false);
```
Comment (curran) [inline packages/neoFrontend/src/pages/VizPage/LinkWithIcon/index.js]: `isHovering` is not used anywhere. I'm not sure why it's here. Please remove this code for `isHovering`, `onMouseEnter`, and `onMouseLeave`. For hover styles, CSS can be used. In any case, for the first pass there is no need for a hovered style.

File: packages/neoFrontend/src/VizPreview/styles.js [hunk 12]
```diff
@@ -5,7 +5,6 @@ import { Footer, Z_WAY_ABOVE } from '../styles';
 export const VizPreviews = styled.div`
   display: flex;
   flex-wrap: wrap;
-  justify-content: center;
```
Comment (curran) [inline packages/neoFrontend/src/VizPreview/styles.js]: This change seems unrelated. Please rebase or otherwise use the latest `VizPreview` from `master`.

File: packages/neoFrontend/src/pages/ProfilePage/Body/index.js [hunk 13]
```diff
@@ -1,22 +1,45 @@
...
 import { Wrapper, Content } from '../../styles';
+import { SidebarWrapper, Main, Sidebar } from '../styles';
+import { LinkWithIcon } from '../../VizPage/LinkWithIcon'
+
 
 export const Body = () => {
   const profilePageData = useContext(ProfilePageDataContext);
   const { user, visualizationInfos } = profilePageData;
 
+  const [visualizations, setVisualizations] = useState(visualizationInfos);
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/Body/index.js]: So the core idea of this sidebar is that you're either viewing only public, or only private, never both. This implementation shows mixed public and private on page load.

Desired behavior:
 * Show public visualizations on page load, with the "Public" link shown in the active style.
 * When clicking "Private", only then show the private visualizations. Ideally there could be a URL param that stores this state, but that can be done in a later PR.

Suggested approach:

```suggestion
  const [activeSideNav, setActiveSideNav] = useState('public');
```

Then filter the `visualizationInfos` behind a `useMemo` while rendering. No need to set the filtered results into a piece of state.

**Review by curran — COMMENTED** [review]
@milafrerichs I've done a thorough review and have dropped my suggestions here.

Please resolve the design conflicts like this (very rough mock, but expresses the desired layout):

image

It may be easiest to branch off the latest master and port the work to there, rather than trying to resolve the merge conflicts.

Good luck! Thanks for your work on this.

File: packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js (hunk 8 above)
Comment (curran) [reply packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js:36]: Also suggest to add a test case: by default only public vizzes should be listed.

Comment (curran) [reply packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js:36]: Also suggest to add a test case: by default the public link should have the active style.

File: packages/neoFrontend/src/pages/ProfilePage/styles.js (hunk 10 above)
Comment (curran) [reply packages/neoFrontend/src/pages/ProfilePage/styles.js]: Please make it work on mobile as well. On mobile, the layout should flow downward like this:

image

File: package.json [hunk 14]
```diff
@@ -13,6 +13,7 @@
     "prettier": "cd packages/neoFrontend && npm run prettier && cd ../neoBackend && npm run prettier"
   },
   "devDependencies": {
+    "@testing-library/jest-dom": "^5.11.1",
```
Comment (curran) [inline package.json]: Why is this needed? What is insufficuent about the Jest setup that's built-in with create-react-app?

**Comment by curran** [comment]
FWIW the design concept is quite similar to GMail's sidebar.

image

File: package.json (hunk 14 above)
Comment (milafrerichs) [reply package.json]: `@testing-library/jest-dom` adds custom matchers like `toBeInTheDocument`

File: packages/neoFrontend/package.json (hunk 1 above)
Comment (milafrerichs) [reply packages/neoFrontend/package.json:54]: testing-library offers great ways of testing the real output of react apps.

File: packages/neoFrontend/src/pages/ProfilePage/Body/index.js [hunk 15]
```diff
@@ -1,21 +1,40 @@
...
 import { ProfilePane } from '../ProfilePane';
-import { ProfilePageDataContext } from '../ProfilePageDataContext';
-import { Wrapper, Content, Centering } from '../../styles';
 import { ProfileMenuBar } from './styles';
 
 export const Body = () => {
   const profilePageData = useContext(ProfilePageDataContext);
   const { user, visualizationInfos } = profilePageData;
 
+  const [privacy, setPrivacy] = useState('public');
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/Body/index.js:24]: Nice.

File: packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+      </AuthContext.Provider>
+    </WarningContext.Provider>
+    </Themed>
+  )
+  render(tree, { wrapper: MemoryRouter })
+  fireEvent.click(screen.getByText('Private'))
+  expect(screen.getByText('My little private visualization')).toBeInTheDocument()
+  expect(screen.queryByText('My big public visualization')).not.toBeInTheDocument()
+})
+test('clicking public shows public viz', () => {
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js:59]: Nice.

File: packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+  const profilePageData = { visualizationInfos: [
+    { id: 1, privacy: 'private', title: 'My little private visualization'},
+    { id: 2, privacy: 'public', title: 'My big public visualization'},
+    { id: 3, privacy: undefined, title: 'My small undefined public visualization'},
```
Comment (curran) [inline packages/neoFrontend/src/pages/ProfilePage/profilepage.test.js:63]: Great work covering this case.

**Review by curran — COMMENTED** [review]
Fantastic! Will merge and refine.
